include/cuco/detail/trie/bit_vector/bit_vector.cuh

```diff
@@ -0,0 +1,200 @@
+   *
+   * @return The base rank
+   */
+  __host__ __device__ uint64_t abs() const
+  {
+    return (static_cast<uint64_t>(abs_hi_) << 8) | abs_lo_;
+  }
+
+  /**
+   * @brief Sets base rank of current 256-bit interval
+   *
+   * @param abs Base rank
+   */
+  void set_abs(uint64_t abs)
+  {
+    abs_hi_ = static_cast<uint32_t>(abs >> 8);
+    abs_lo_ = static_cast<uint8_t>(abs);
+  }
+};
+
+/**
+ * @brief Union of 64-bit word with rank
+ *
+ * Need this so that all aow_storage structures in bitvector have 64-bit element type
+ */
+union rank_union {
+  uint64_t word_;  ///< word view
+  rank rank_;      ///< rank view
+};
+
+/**
+ * @brief Bitvector class with rank and select index structures
+ *
+ * In addition to standard bitvector get/set operations, this class provides
+ * rank and select operation API. It maintains index structures to make both these
+ * new operations close to constant time.
+ * Bitvector construction happens on host, after which the structures are moved to device.
+ * All subsequent read-only operations access device structures only.
+ *
+ * @tparam Extent Data structure size type
+ * @tparam Scope The scope in which operations will be performed by individual threads.
+ * @tparam Allocator Type of allocator used for device storage
+ * @tparam Storage Slot window storage type
+ */
+
+template <class Extent             = cuco::experimental::extent<std::size_t>,
+          cuda::thread_scope Scope = cuda::thread_scope_device,
+          class Allocator          = cuco::cuda_allocator<std::byte>,
+          class Storage            = cuco::experimental::storage<1>>
+class bit_vector {
+ public:
+  bit_vector();
+  ~bit_vector();
+
+  /**
+   * @brief adds a new bit at the end
+   *
+   * Grows internal storage if needed
+   *
+   * @param bit Boolean value of new bit to be added
+   */
+  void append(bool bit);
+
+  /**
+   * @brief Builds indexes for rank and select
+   *
+   * Also creates device-side snapshot
+   */
+  void build();
+
+  using size_type = typename Extent::value_type;  ///< size type to specify bit index
+  /**
+   * @brief Modifies a single bit
+   *
+   * @param index position of bit to be modified
+   * @param bit new value of bit
+   */
+  void set(size_type index, bool bit);
+
+  /**
+   * @brief Sets last bit to specified value
+   *
+   * @param bit new value of last bit
+   */
+  void set_last(bool bit);
+
+  static constexpr auto cg_size      = 1;      ///< CG size used for probing
+  static constexpr auto window_size  = 1;      ///< Window size used for probing
+  static constexpr auto thread_scope = Scope;  ///< CUDA thread scope
+
+  using extent_type =
+    decltype(make_window_extent<cg_size, window_size>(std::declval<Extent>()));  ///< Extent type
+  using allocator_type = Allocator;                                              ///< Allocator type
+  using storage_type =
+    detail::storage<Storage, size_type, extent_type, allocator_type>;  ///< Storage type
```

Comment:
```suggestion
  using slot_type = uint64_t;
  using storage_type =
    aow_storage<slot_type, 1, size_type, allocator_type>;  ///< Storage type
```
The slot type should be independent of `size_type`. Imagine `using size_type = int32_t`;
Also, can use `aow_storage` directly to simplify the logic.

Reply:
d063cc50bc0b5ad0b02b8f032ba75c4a55b2f968